verification/certifyKey_test.go

```diff
@@ -118,21 +579,194 @@ func testCertifyKey(d TestDPEInstance, t *testing.T) {
+
+		// TODO: When DeriveChild is implemented, call it here to add more TCIs and call CertifyKey again.
+	}
+}
+
+// Build certificate chain and calls to validateSignature on each chain.
+func validateCertChain(t *testing.T, certChain []*x509.Certificate, leafCert *x509.Certificate) {
+	t.Helper()
+	var certsToProcess []*x509.Certificate
+	if leafCert != nil {
+		t.Log("[LOG]: Validating leaf certificate chain...")
+		certsToProcess = []*x509.Certificate{leafCert}
+	} else {
+		t.Log("[LOG]: Validating intermediate certificates chains...")
+		certsToProcess = certChain
+	}
+
+	// Remove unhandled critical extensions reported by x509 but defined in spec
+	t.Log("[LOG]: Checking for unhandled critical certificate extensions unknown to DPE certificates profile spec...")
+	removeTcgDiceCriticalExtensions(t, certsToProcess)
+
+	// Remove unhandled extended key usages reported by x509 but defined in spec
+	t.Log("[LOG]: Checking for extended key usages unknown to DPE certificates profile spec...")
+	removeTcgDiceExtendedKeyUsages(t, certsToProcess)
+
+	// Build verify options
+	opts := buildVerifyOptions(t, certChain)
+
+	if leafCert != nil {
+		// Certificate chain validation for leaf
+		chains, err := leafCert.Verify(opts)
+		if err != nil {
+			// Certificate chain cannot be built from leaf to root
+			t.Errorf("[ERROR]: Error in certificate chain %s: ", err.Error())
+		}
+
+		// Log certificate chains linked to leaf
+		t.Logf("[LOG]: Chains of DPE leaf certificate.")
+		for _, chain := range chains {
+			for i, cert := range chain {
+				t.Logf("%d %s", i, (*cert).Subject)
+			}
+
+		}
+
+		// This indicates that signature validation found no errors in the DPE leaf cert chain
+		t.Logf("[LOG]: DPE leaf certificate chain validation is done")
+
+	} else {
+		// Certificate chain validation for each intermediate certificate
+		for _, cert := range certChain {
+			chains, err := cert.Verify(opts)
+			if err != nil {
+				t.Errorf("[ERROR]: Error in Certificate Chain of %s: %s", cert.Subject, err.Error())
+			}
+
+			// Log certificate chains linked to each cetificate in chain
+			t.Logf("[LOG]: Chains of intermediate certificate.")
+			for _, chain := range chains {
+				for i, cert := range chain {
+					t.Logf("%d %s", i, (*cert).Subject)
+				}
+			}
+		}
+
+		// This indicates that signature validation found no errors each cert
+		// chain of intermediate certificates
+		t.Logf("[LOG]: Intermediate certificates chain validation is done")
+	}
+}
+
+func buildVerifyOptions(t *testing.T, certChain []*x509.Certificate) x509.VerifyOptions {
+	roots := x509.NewCertPool()
+	intermediates := x509.NewCertPool()
+
+	t.Logf("[LOG]: Root certificate is expected to be in the beginning of the chain, the rest are expected to be intermediates.")
+	if certChain[0].Subject.String() != certChain[0].Issuer.String() {
+		t.Errorf("[ERROR]: Found a non-root certificate in beginning of certificate chain returned by GetCertificateChain.")
+	} else {
```

Comment:
It is not a requirement that the root is self-signed. That just happens to be true for the current targets.

Reply:
This check is removed.